grovel/grovel.lisp

```diff
@@ -767,22 +885,80 @@ string."
 (defun cc-include-grovel-argument ()
   (format nil "-I~A" (truename (system-source-directory :cffi-grovel))))
 
+(defun process-from-cache-p (system cached-lisp-file cached-lib-file)
+  (and cached-lib-file
+       cached-lisp-file
+       (uiop:file-exists-p (system-relative-pathname system cached-lib-file))
+       (uiop:file-exists-p (system-relative-pathname system cached-lisp-file))))
+
 ;;; *PACKAGE* is rebound so that the IN-PACKAGE form can set it during
 ;;; *the extent of a given wrapper file.
-(defun process-wrapper-file (input-file
-                             &key
-                               (output-defaults (make-pathname :defaults input-file :type "processed"))
-                               lib-soname)
+(defun process-wrapper-file (system input-file dest-lisp-file dest-lib-file c-file
+                             o-file &key lib-soname cache-dir)
   (with-standard-io-syntax
-    (multiple-value-bind (c-file lisp-forms)
-        (generate-c-lib-file input-file output-defaults)
-    (let ((lib-file (make-lib-file-name (make-soname lib-soname output-defaults)))
-          (o-file (make-o-file-name output-defaults "__wrapper")))
-        (cc-compile o-file (list (cc-include-grovel-argument) c-file))
-        (link-shared-library lib-file (list o-file))
-        ;; FIXME: hardcoded library path.
-        (values (generate-bindings-file lib-file lib-soname lisp-forms output-defaults)
-                lib-file)))))
+    (multiple-value-bind (input-data feature-expressions)
+        (read-wrapper-spec input-file)
+      (let* ((cached-lib-file (feature-specific-cache-file
+                               dest-lib-file cache-dir feature-expressions))
+             (cached-lisp-file (feature-specific-cache-file
+                                dest-lisp-file cache-dir feature-expressions))
+             (*local-includes* nil))
+        (if (process-from-cache-p system cached-lisp-file cached-lib-file)
+            (process-wrapper-file-from-cache system dest-lib-file dest-lisp-file
+                                             c-file o-file
+                                             cached-lib-file cached-lisp-file)
+            (process-wrapper-file-from-scratch system input-data dest-lisp-file
+                                               lib-soname dest-lib-file c-file
+                                               o-file cached-lib-file
+                                               cached-lisp-file))
+        (values dest-lisp-file dest-lib-file)))))
```

Comment:
As with the groveller version higher up in this file, this needs to check if we have the files cached so it knows if it needs to build them ('from-scratch') or if it can just copy them ('from-cache')